pip/main.ts

```diff
@@ -0,0 +1,158 @@
+import {
+  green,
+  red,
+  yellow,
+} from "https://deno.land/std@0.150.0/fmt/colors.ts";
+import { PhylumApi } from "phylum";
+
+// Find project root directory.
+async function findRoot(manifest: string): Promise<string | undefined> {
+  let workingDir = Deno.cwd();
+
+  // Traverse up to 32 directories to find the root directory.
+  for (let i = 0; i < 32; i++) {
+    try {
+      // Check if manifest exists at location.
+      await Deno.stat(workingDir + "/" + manifest);
+      return workingDir;
+    } catch (_e) {
+      // Pop to parent if manifest doesn't exist.
+      workingDir += "/..";
+    }
+  }
+
+  return undefined;
+}
+
+// List with all of pip's subcommands.
+const knownSubcommands = [
+  "install",
+  "download",
+  "uninstall",
+  "freeze",
+  "inspect",
+  "list",
+  "show",
+  "check",
+  "config",
+  "search",
+  "cache",
+  "index",
+  "wheel",
+  "hash",
+  "completion",
+  "debug",
+  "help",
+];
+
+// Ensure the first argument is a known subcommand.
+//
+// This prevents us from skipping the analysis when an argument is passed before
+// the first subcommand (i.e.: `pip --no-color install package`).
+const subcommand = Deno.args[0];
+if (Deno.args.length != 0 && !knownSubcommands.includes(subcommand)) {
+  console.error(`[${red("phylum")}] This extension does not support arguments before the first subcommand. Please open an issue if "${subcommand}" is not an argument.`);
+}
+
+// Ignore all commands that shouldn't be intercepted.
+if (Deno.args.length == 0 || subcommand != "install") {
+  const cmd = Deno.run({ cmd: ["pip", ...Deno.args] });
+  const status = await cmd.status();
+  Deno.exit(status.code);
+}
+
+// Analyze new dependencies with phylum before install/update.
+await checkDryRun();
+
+// Perform the package installation.
+const cmd = Deno.run({ cmd: ["pip", ...Deno.args] });
+const status = await cmd.status();
+Deno.exit(status.code);
+
+// Analyze new packages.
+async function checkDryRun() {
+  console.log(`[${green("phylum")}] Finding new dependencies…`);
+
+  const status = PhylumApi.runSandboxed({
+    cmd: "pip",
+    args: [...Deno.args, "--dry-run"],
+    exceptions: {
+      run: ["/bin"],
```

Comment:
This is a placeholder. Going to figure out what to do about Python once I know about the macOS sandboxing requirements.